crates/aptos-telemetry/src/sender.rs

```diff
@@ -0,0 +1,224 @@
+                        debug!("Retrying failed send telemetry event: Error: {}.", error);
+                        continue;
+                    }
+                    error!("Failed to send telemetry event: Error: {}", error);
+                    metrics::increment_telemetry_service_failures(&event_name);
+                    return;
+                }
+            }
+        }
+    }
+
+    async fn post_metrics(&self, telemetry_dump: &TelemetryDump) -> Result<(), anyhow::Error> {
+        let token = self.get_token().await?;
+
+        // Send the request and wait for a response
+        let send_result = self
+            .client
+            .post(METRICS_URL)
+            .json::<TelemetryDump>(telemetry_dump)
+            .bearer_auth(token)
+            .send()
+            .await;
+
+        // Process the response
+        match send_result {
+            Ok(response) => {
+                let status_code = response.status();
+                if status_code.is_success() {
+                    Ok(())
+                } else if status_code == StatusCode::UNAUTHORIZED {
+                    self.reset_token();
+                    Err(anyhow!("Unauthorized"))
+                } else {
+                    Err(anyhow!("Error status received: {}", status_code))
+                }
+            }
+            Err(error) => Err(anyhow!("Error sending metrics. Err: {}", error)),
+        }
+    }
+
+    async fn get_token(&self) -> Result<String, Error> {
+        // Try to read the token holding a read lock
+        let token = { self.auth_context.token.read().as_ref().cloned() };
+        match token {
+            Some(token) => Ok(token),
+            None => {
+                let token = self.authenticate().await?;
+                *self.auth_context.token.write() = Some(token.clone());
+                Ok(token)
+            }
+        }
+    }
+
+    async fn get_public_key_from_server(&self) -> Result<x25519::PublicKey, anyhow::Error> {
+        let url = Url::parse(METRICS_ENDPOINT)?;
+        let response = self.client.get(url).send().await?;
+
+        match response.error_for_status() {
+            Ok(response) => {
+                let public_key = response.json::<x25519::PublicKey>().await?;
+                Ok(public_key)
+            }
+            Err(err) => Err(anyhow!("Error getting server public key. {}", err)),
+        }
+    }
+
+    async fn server_public_key(&self) -> Result<x25519::PublicKey, anyhow::Error> {
+        let server_public_key = { *self.auth_context.server_public_key.lock() };
+        match server_public_key {
+            Some(key) => Ok(key),
+            None => {
+                let public_key = self.get_public_key_from_server().await?;
+                *self.auth_context.server_public_key.lock() = Some(public_key);
+                Ok(public_key)
+            }
+        }
+    }
+
+    fn reset_token(&self) {
+        *self.auth_context.token.write() = None
```

Comment:
Instead of resetting this to `None`, should you generate a new valid token instead?

Reply:
It might make the control flow hard to follow. Right now we have a flow like: get_token() -> authenticate() -> get_public_key() -> post metrics -> reset_token(). If I refresh the token here, then we will have two place where authentication happens - once in get_token and once in reset_token - and we will need to handle errors for reset_token as well. We anyhow wrap all this with a retry mechanism, so when there is a retry, the token will be refreshed.